lib/ecto/association.ex

```diff
@@ -217,14 +218,14 @@ defmodule Ecto.Association do
   @doc """
   Add the default assoc query where clauses a provided query
   """
-  def combine_assoc_query(assoc_queryable, nil), do: assoc_queryable
-  def combine_assoc_query(%Ecto.Query{wheres: assoc_wheres}, provided_queryable) do
-    provided_queryable
-    |> Ecto.Queryable.to_query()
-    |> Map.update!(:wheres, &Kernel.++(&1, assoc_wheres))
+  def combine_assoc_query(%{queryable: queryable} = assoc, nil), do: combine_assoc_query(assoc, queryable)
+  def combine_assoc_query(%{where: nil}, queryable), do: queryable
+  def combine_assoc_query(%{where: []}, queryable), do: queryable
+  def combine_assoc_query(%{where: {module, function, args}} = assoc, queryable) do
+    combine_assoc_query(%{assoc | where: apply(module, function, args)}, queryable)
```

Comment:
I think we should call Ecto.Query.where with the result of `apply`. Otherwise we will accidentally support returning `nil` or another MFA from the callback.

Reply:
Isn't that what this eventually does? This just updates the `where` and recurses the next clause that does call `Ecto.Query.where/3`

Reply:
Ah, I see your other feedback to directly call where. I don't mind that at all :) Will add that hange.

Reply:
Exactly. We don't want to recurse. :)